### ISSUE TITLE
Regression since Meteor 1.7.x: Order of magnitude more "fiber" objects created

### ISSUE THREAD

**Issue by ebbe-brandstrup** [meteor/meteor#10359]
Meteor version: since 1.7.x
OS: Ubuntu 16.04 and OS X

### TL;DR

Fibers accumulate slowly over time with Meteor 1.6, but extremely fast on Meteor 1.7+. After a while, the app is completely unresponsive (100% CPU) and must be restarted.

### Longer story

When running a lot of operations that use fibers, Meteor 1.7+ creates millions of fiber objects that are never released.

We had to rollback to Meteor 1.6 in production because our app hosed its heap with these fibers. After 1-2 days of normal use (with 10-20 concurrent users in a complex app) the app would get stuck using 100% CPU and was completely unresponsive. We could not even inspect it with node inspect. We have to restart the app when this occurs. Doing heap snapshots before the unresponsive state occurs reveals the many fiber objects:

<img width="1440" alt="fibers leak on meteor 1 7 in production" src="https://user-images.githubusercontent.com/4715559/49138891-dcd4f800-f309-11e8-8028-d77aa45a64d3.png">

The "leak" of fiber objects also occurs with Meteor 1.6, but at a much slower pace. We get stuck in the same unresponsive state roughly once a month.

**Expected behavior**
Something related to creation of fibers (via [meteor-promise](https://github.com/meteor/promise)) changed from Meteor 1.6 to Meteor 1.7+. I would expect the number of fibers created for identical apps should be similar between Meteor 1.6 and later versions.

All those versions of Meteor use the same meteor-promise@0.8.6 and fibers@2.0.0, so I have ruled out those repos as sources of the regression.

**Actual behaviour**
Lots more fibers are created when using Meteor 1.7+.

**Steps to reproduce**
I have created a simple example repo based on Meteor 1.6.1.4 [here](https://github.com/ebbe-brandstrup/meteor-1.6-fibers-regression-reproduction):

1.
```
git clone https://github.com/ebbe-brandstrup/meteor-1.6-fibers-regression-reproduction
cd meteor-1.6-fibers-regression-reproduction
meteor npm install
```

2.
Then `meteor npm run start` or `meteor npm run start:gc` to see GC output as well.

3.
Then open some browser tabs on http://localhost:3000 and observe the server console output report the number of fibers created.

Opening 5+ tabs (and having them visible on-screen to prevent your browser from throttling JS) will cause much fewer fibers to be created than in the corresponding Meteor 1.8.0.1 version of the example app that I created [here](https://github.com/ebbe-brandstrup/meteor-1.8-fibers-regression-reproduction).

Follow the same steps as the 1.6 version and observe that the number of fibers created is ~x10 higher (to compare, be sure to have the same number of tabs open on `http://localhost:3000` and reload them if restarting the server).

Please note that the example is very contrived. Every connected browser calls a Meteor method every 50ms which does 500 `findOne` queries in a collection with 1 doc. That's obviously not a representative "real" app, but it demonstrates the difference in the number of created fibers in those versions of Meteor.

#### Meteor 1.6.1.4

When running the Meteor 1.6.1.4 version, this is the heap snapshot after running with 8 browser clients for ~2 minutes:

<img width="932" alt="heap snapshot" src="https://user-images.githubusercontent.com/4715559/49141859-06455200-f311-11e8-9fb0-5304b9133668.png">

and the corresponding server console output:

![console output](https://user-images.githubusercontent.com/4715559/49141905-1fe69980-f311-11e8-8085-4f37b72661d9.png)

And an example of the allocation stack of a fiber:

<img width="936" alt="allocation stack example" src="https://user-images.githubusercontent.com/4715559/49141999-486e9380-f311-11e8-9df4-f1a9fdeef606.png">

#### Meteor 1.8.0.1

Comparing those to the Meteor 1.8.0.1 version there are some noticeable differences (heap snapshot after ~2 minutes also with 8 browser clients; note the number of fibers in the heap is larger than the `fibersCreated` reported from the `fibers` package shown in the server console output):

<img width="764" alt="heap snapshot" src="https://user-images.githubusercontent.com/4715559/49142569-828c6500-f312-11e8-939b-e2de64fc61ef.png">

Server console output from the same run:

<img width="1090" alt="console output" src="https://user-images.githubusercontent.com/4715559/49142598-959f3500-f312-11e8-91e8-a02d56319ecf.png">

And different allocation stack of a similar fiber:

<img width="935" alt="allocation stack example" src="https://user-images.githubusercontent.com/4715559/49142655-c41d1000-f312-11e8-8490-e8c3967d4c22.png">

Something to note is that the fibers in the Meteor 1.6 version are all attached to the root GC object (they have a distance other than '-' show in the Chrome DevTools). However, in the Meteor 1.8 version, most of the fibers have a distance of '-' shown:

<img width="1185" alt="heap snapshot - retainer of fiber with no distance" src="https://user-images.githubusercontent.com/4715559/49142763-fcbce980-f312-11e8-890e-6f70be70fcd8.png">

And jumping to the retainer of that fiber:

<img width="1116" alt="heap snapshot - expanded retainer of fiber with no distance" src="https://user-images.githubusercontent.com/4715559/49142809-1100e680-f313-11e8-9600-865eae00a157.png">

Maybe these differences will make the cause of this regression easier to discern.

### Possibly related issues

* Old GC CPU fiber bomb issue that does a good job of explaining the CPU stalling: https://github.com/meteor/meteor/issues/7747
* `meteor-promise` fiber pool issue: https://github.com/meteor/promise/issues/14

**Comment by KoenLav** [comment]
Judging from the allocation stack of the Fibers this is the place where the enormous amount of Fibers are created: https://github.com/meteor/meteor/blob/devel/packages/mongo/doc_fetcher.js

When Meteor can't use the oplog for a query, if places a call to DocFetcher to get the document. These calls however aren't queued/batched even when we can reasonably assume that the result of these calls is to be passed to a single callback (either as an array or as singular invocations).

This is something I also noticed when reworking the entire Oplog => DDP mechanism. I hope these extensive insights will help raise some more concern amongst MDG...

**Comment by VeselyT** [comment]
We are also experiencing this issue on one of our applications. Thanks to @ebbe-brandstrup reproduction example we were able to bisect to commit https://github.com/meteor/meteor/commit/afb7d23d045602cd7f9d4b7ae61081c4bcc3b12f that seems to cause this regression. This commit introduced use of promises inside meteor mongo driver. 

Problem seems that Fibers used inside promises are not properly released inside fiber pool (for fibers that exceed size of the pool... with default value 20 it's easy to exceed). 

When we reset handle to Fiber after we are done with it it seems to fix this issue https://github.com/ExentriqLtd/promise/commit/bbe4f0d20b70417950381aea112993c4cc8c1168 (atleast for this reproduction case.. there is still increase of fibers apparent but they all seems to be released when those method calls stop and all methods results are delivered over DDP)

We will see If this will fix our issue on production or not in few days (as our app gets stucked not very regularly... sometimes it runs fine for few days without problem and sometimes we have couple of issues in one day). However, in the meantime you can try it to see if it solves your issue. 

We hope this will help to fix this issue or at least it will point someone in the right direction.

**Comment by KoenLav** [comment]
@VeselyT great work!

@laverdet is this an erroneous implementation of Fibers, or something which may be fixable within your node-fibers package?

**Comment by ebbe-brandstrup** [comment]
Thanks and very interesting!

I am trying out [your meteor-promise commit](https://github.com/ExentriqLtd/promise/commit/bbe4f0d20b70417950381aea112993c4cc8c1168) to test if it solves our issue.

**Comment by benjamn** [comment]
Ok, I've released a new version of the Meteor `promise` package that uses version 0.8.7 of the `meteor-promise` npm package, which includes @VeselyT's commit. In other words, running
```bash
meteor update promise
```
may begin to help with this issue, though it's possible that the copy of the `promise` package that's actually responsible for this leak is embedded as a dependency of the `meteor` command-line tool, which means it might not be fixed until the next release of Meteor. If you're willing to try a beta release of Meteor 1.8.1, just watch #10248 for version 1.8.1-beta.8.

**Comment by benjamn** [comment]
I can confirm that Fiber allocation in a Meteor 1.8.1-beta.8 app looks a lot more like the Meteor 1.6 profile @ebbe-brandstrup provided above. Please give it a try!

**Comment by wildhart** [comment]
I'm not sure if this is the same bug, but I'm also getting a serious memory leak since upgrading from Meteor 1.6.1.1 to Meteor 1.8.  Monitoring my production server shows a steady 15-20MB per hour RAM increase with a steady ~80 sessions.  My fibre logs (logged every 10 minutes) don't increase as rapidly as above though, but still growing as a polynomial `fibres = 0.1572min² + 89.199min + 599.66 (r²=0.9999)`:
````
06T10:52:51 elapsed: 0m rss: 154mb heapTotal: 114mb heapUsed: 78mb external: 25mb fibers: 5
06T11:02:51 elapsed: 10m rss: 203mb heapTotal: 149mb heapUsed: 114mb external: 45mb fibers: 1242
06T11:12:51 elapsed: 20m rss: 203mb heapTotal: 148mb heapUsed: 103mb external: 45mb fibers: 2344
06T11:22:51 elapsed: 30m rss: 205mb heapTotal: 150mb heapUsed: 107mb external: 45mb fibers: 3421
06T11:32:51 elapsed: 40m rss: 206mb heapTotal: 151mb heapUsed: 116mb external: 45mb fibers: 4505
06T11:42:51 elapsed: 50m rss: 209mb heapTotal: 152mb heapUsed: 113mb external: 45mb fibers: 5662
06T11:52:51 elapsed: 60m rss: 208mb heapTotal: 152mb heapUsed: 110mb external: 45mb fibers: 6808
06T12:02:51 elapsed: 70m rss: 209mb heapTotal: 153mb heapUsed: 108mb external: 45mb fibers: 7901
06T12:12:51 elapsed: 80m rss: 211mb heapTotal: 153mb heapUsed: 109mb external: 45mb fibers: 8992
06T12:22:51 elapsed: 90m rss: 213mb heapTotal: 155mb heapUsed: 112mb external: 45mb fibers: 10136
06T12:32:51 elapsed: 100m rss: 214mb heapTotal: 157mb heapUsed: 114mb external: 45mb fibers: 11332
06T12:42:51 elapsed: 110m rss: 216mb heapTotal: 157mb heapUsed: 116mb external: 45mb fibers: 12533
06T12:52:51 elapsed: 120m rss: 216mb heapTotal: 156mb heapUsed: 113mb external: 45mb fibers: 13784
06T13:02:51 elapsed: 130m rss: 217mb heapTotal: 158mb heapUsed: 116mb external: 46mb fibers: 15045
06T13:12:51 elapsed: 140m rss: 222mb heapTotal: 161mb heapUsed: 114mb external: 45mb fibers: 16297
06T13:22:51 elapsed: 150m rss: 225mb heapTotal: 163mb heapUsed: 121mb external: 45mb fibers: 17658
06T13:32:51 elapsed: 160m rss: 228mb heapTotal: 165mb heapUsed: 118mb external: 45mb fibers: 19122
06T13:42:51 elapsed: 170m rss: 227mb heapTotal: 164mb heapUsed: 121mb external: 45mb fibers: 20513
06T13:52:51 elapsed: 180m rss: 229mb heapTotal: 166mb heapUsed: 122mb external: 45mb fibers: 21871
06T14:02:51 elapsed: 190m rss: 233mb heapTotal: 169mb heapUsed: 114mb external: 45mb fibers: 23190
06T14:12:51 elapsed: 200m rss: 232mb heapTotal: 168mb heapUsed: 120mb external: 45mb fibers: 24568
06T14:22:51 elapsed: 210m rss: 233mb heapTotal: 168mb heapUsed: 117mb external: 45mb fibers: 25908
06T14:32:51 elapsed: 220m rss: 235mb heapTotal: 169mb heapUsed: 118mb external: 45mb fibers: 27273
06T14:42:51 elapsed: 230m rss: 238mb heapTotal: 172mb heapUsed: 130mb external: 45mb fibers: 28662
06T14:52:51 elapsed: 240m rss: 243mb heapTotal: 175mb heapUsed: 118mb external: 45mb fibers: 30263
06T15:02:51 elapsed: 250m rss: 249mb heapTotal: 181mb heapUsed: 130mb external: 45mb fibers: 32119
06T15:12:51 elapsed: 260m rss: 254mb heapTotal: 183mb heapUsed: 130mb external: 46mb fibers: 33992
06T15:22:51 elapsed: 270m rss: 256mb heapTotal: 185mb heapUsed: 122mb external: 45mb fibers: 35898
06T15:32:51 elapsed: 280m rss: 258mb heapTotal: 186mb heapUsed: 135mb external: 46mb fibers: 37635
06T15:42:51 elapsed: 290m rss: 258mb heapTotal: 185mb heapUsed: 125mb external: 45mb fibers: 39413
06T15:52:51 elapsed: 300m rss: 262mb heapTotal: 189mb heapUsed: 129mb external: 46mb fibers: 41363
06T16:02:51 elapsed: 310m rss: 265mb heapTotal: 191mb heapUsed: 135mb external: 46mb fibers: 43430
06T16:12:51 elapsed: 320m rss: 268mb heapTotal: 193mb heapUsed: 135mb external: 52mb fibers: 45301
06T16:22:51 elapsed: 330m rss: 271mb heapTotal: 194mb heapUsed: 131mb external: 46mb fibers: 47412
06T16:32:51 elapsed: 340m rss: 277mb heapTotal: 198mb heapUsed: 134mb external: 46mb fibers: 49512
06T16:42:51 elapsed: 350m rss: 275mb heapTotal: 197mb heapUsed: 132mb external: 46mb fibers: 51431
06T16:52:51 elapsed: 360m rss: 277mb heapTotal: 199mb heapUsed: 133mb external: 56mb fibers: 53421
06T17:02:51 elapsed: 370m rss: 280mb heapTotal: 201mb heapUsed: 141mb external: 46mb fibers: 55537
06T17:12:51 elapsed: 380m rss: 284mb heapTotal: 203mb heapUsed: 131mb external: 46mb fibers: 57596
06T17:22:51 elapsed: 390m rss: 287mb heapTotal: 205mb heapUsed: 130mb external: 46mb fibers: 59650
06T17:32:51 elapsed: 400m rss: 290mb heapTotal: 207mb heapUsed: 141mb external: 46mb fibers: 61822
06T17:42:51 elapsed: 410m rss: 290mb heapTotal: 209mb heapUsed: 142mb external: 46mb fibers: 63820
06T17:52:51 elapsed: 420m rss: 288mb heapTotal: 207mb heapUsed: 133mb external: 46mb fibers: 65955
06T18:02:51 elapsed: 430m rss: 293mb heapTotal: 210mb heapUsed: 151mb external: 46mb fibers: 68347
06T18:12:51 elapsed: 440m rss: 296mb heapTotal: 211mb heapUsed: 143mb external: 46mb fibers: 70631
06T18:22:51 elapsed: 450m rss: 298mb heapTotal: 213mb heapUsed: 145mb external: 46mb fibers: 72784
06T18:32:51 elapsed: 460m rss: 300mb heapTotal: 214mb heapUsed: 138mb external: 46mb fibers: 74967
06T18:42:51 elapsed: 470m rss: 302mb heapTotal: 215mb heapUsed: 139mb external: 46mb fibers: 77170
06T18:52:51 elapsed: 480m rss: 306mb heapTotal: 218mb heapUsed: 153mb external: 46mb fibers: 79376
06T19:02:51 elapsed: 490m rss: 309mb heapTotal: 220mb heapUsed: 140mb external: 46mb fibers: 81687
06T19:12:51 elapsed: 500m rss: 312mb heapTotal: 221mb heapUsed: 151mb external: 46mb fibers: 83998
06T19:22:51 elapsed: 510m rss: 316mb heapTotal: 224mb heapUsed: 149mb external: 46mb fibers: 86537
06T19:32:51 elapsed: 520m rss: 321mb heapTotal: 227mb heapUsed: 152mb external: 46mb fibers: 88926
06T19:42:51 elapsed: 530m rss: 326mb heapTotal: 230mb heapUsed: 161mb external: 47mb fibers: 91619
06T19:52:51 elapsed: 540m rss: 332mb heapTotal: 233mb heapUsed: 150mb external: 46mb fibers: 94510
06T20:02:51 elapsed: 550m rss: 337mb heapTotal: 238mb heapUsed: 151mb external: 46mb fibers: 97590
06T20:12:51 elapsed: 560m rss: 338mb heapTotal: 237mb heapUsed: 156mb external: 48mb fibers: 100454
````
I haven't been able to take any production heapdumps yet because I can't get `node-heapdump` to work on my production server (`Error: The module '/built_app/programs/server/npm/node_modules/heapdump/build/Release/addon.node'
was compiled against a different Node.js version using NODE_MODULE_VERSION 46. This version of Node.js requires NODE_MODULE_VERSION 57...`).

Is it safe to try `Meteor 1.8.1-beta.8` in production?

**Comment by benjamn** [comment]
@wildhart While I believe Meteor 1.8.1 is safe to use in production at this point, your personal risk tolerance may disagree. Maybe try doing `meteor update promise` first?

**Comment by wildhart** [comment]
`meteor update promise` seems to have fixed it for me, updating from 0.11.1 to 0.11.2.  Redeployed at 14:53 with a memory leak before then (despite dropping sessions), and steady after:

![image](https://user-images.githubusercontent.com/12505024/49631876-99d8fa00-fa58-11e8-9568-a9b962dba1f1.png)

Thanks!

**Comment by jamesmillerburgess** [comment]
Our team also noticed a memory leak since launching our app last week, and we were about to start investigating. 

Thanks for saving us a lot of time! 🎉

**Comment by benjamn** [comment]
Yay, glad to hear the `promise` package update was sufficient. That means we can move this issue from [the Release 1.8.1 milestone](https://github.com/meteor/meteor/milestone/61) to [Package Patches](https://github.com/meteor/meteor/milestone/40).

Closing this issue for now, though we can reopen it if the same problem (emphasis on *same*) persists for other folks.

**Comment by KoenLav** [comment]
@VeselyT could you confirm the problem you fixed is similar to the problem described here: https://github.com/laverdet/node-fibers/pull/232

**Comment by VeselyT** [comment]
@KoenLav yes, it seems same to me

**Comment by KoenLav** [comment]
@laverdet @benjamn given the fact that the issue  reported in the node-fibers repository and the solutions raised look so similar it might be worthwhile to investigate whether the PR raised in node-fibers also alleviates the memory leak.

This could be beneficial to the larger node community.
